test-data/unit/check-typeddict.test

```diff
@@ -2122,3 +2122,138 @@ from mypy_extensions import TypedDict
 Foo = TypedDict('Foo', {'camelCaseKey': str})
 value: Foo = {}  # E: Missing key "camelCaseKey" for TypedDict "Foo"
 [builtins fixtures/dict.pyi]
+
+
+[case testTypedDictUnionGetItem]
+from typing import TypedDict, Union
+
+class Foo1(TypedDict):
+    z: str
+    a: int
+class Foo2(TypedDict):
+    z: str
+    b: int
+
+def func(foo: Union[Foo1, Foo2]) -> str:
+    reveal_type(foo["z"])  # ok
+
+    reveal_type(foo["a"])
+    reveal_type(foo["b"])
+    reveal_type(foo["missing"])
+    reveal_type(foo[1])
+
+    return foo["z"]
+[out]
+main:11: note: Revealed type is "builtins.str"
+main:13: note: Revealed type is "Union[builtins.int, Any]"
+main:13: error: TypedDict "Foo2" has no key "a"
+main:14: note: Revealed type is "Union[Any, builtins.int]"
+main:14: error: TypedDict "Foo1" has no key "b"
+main:15: note: Revealed type is "Any"
+main:15: error: TypedDict "Foo1" has no key "missing"
+main:15: error: TypedDict "Foo2" has no key "missing"
+main:16: note: Revealed type is "Any"
+main:16: error: TypedDict key must be a string literal; expected one of ("z", "a")
+main:16: error: TypedDict key must be a string literal; expected one of ("z", "b")
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+
+[case testTypedDictUnionSetItem]
+from typing import TypedDict, Union
+
+class Foo1(TypedDict):
+    z: str
+    a: int
+class Foo2(TypedDict):
+    z: str
+    b: int
+
+def func(foo: Union[Foo1, Foo2]):
+    foo["z"] = "a"  # ok
+    foo["z"] = 1
+
+    foo["a"] = 1
+    foo["b"] = 2
+
+    foo["missing"] = 1
+    foo[1] = "m"
```

Comment:
Can you also try directly calling `foo.__setitem__` (e.g. `foo.__setitem__("z", "a")`)? It's not a big deal if this doesn't work as expected, but it should at least not crash.